src/InsertTagHandler.php

```diff
@@ -0,0 +1,117 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of terminal42/contao-inserttags.
+ *
+ * (c) terminal42
+ *
+ * @license MIT
+ */
+
+namespace Terminal42\InsertTagsBundle;
+
+use Contao\Database;
+use Contao\FrontendUser;
+use Contao\PageModel;
+use Contao\StringUtil;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
+
+class InsertTagHandler
+{
+    private RequestStack $requestStack;
+    private TokenStorageInterface $tokenStorage;
+
+    public function __construct(RequestStack $requestStack, TokenStorageInterface $tokenStorage)
+    {
+        $this->requestStack = $requestStack;
+        $this->tokenStorage = $tokenStorage;
+    }
+
+    /**
+     * Check if a tag should be applied (rules, pages).
+     */
+    public function isTagValid(array $record): bool
+    {
+        // Show for guests only, but member logged in
+        if ($record['guests'] && ($user = $this->getFrontendUser()) !== null) {
+            return false;
+        }
+
+        // Show to members only
+        if ($record['protected']) {
+            $user = $user ?? $this->getFrontendUser();
+
+            // no member logged in
+            if (null === $user) {
+                return false;
+            }
+
+            $tagGroups = StringUtil::deserialize($record['groups']);
+            $userGroups = StringUtil::deserialize($user->groups);
```

Comment:
Did you check the new "guest" group logic in Contao?

Reply:
Should be there :+1: